curvefs/src/client/volume/extent_cache.h

```diff
@@ -0,0 +1,125 @@
+
+#include <cstring>
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+#include "curvefs/proto/metaserver.pb.h"
+#include "curvefs/src/client/volume/extent.h"
+#include "curvefs/src/volume/common.h"
+#include "src/common/concurrent/rw_lock.h"
+
+namespace curvefs {
+namespace client {
+
+using ::curvefs::volume::ReadPart;
+using ::curvefs::volume::WritePart;
+
+struct AllocPart {
+    ExtentAllocInfo allocInfo;
+    size_t writelength = 0;
+    const char* data = nullptr;
+};
+
+struct ExtentCacheOption {
+    uint64_t preallocSize = 64ULL * 1024;
+    uint64_t rangeSize = 1ULL * 1024 * 1024 * 1024;
+};
+
+class ExtentCache {
+ public:
+    ExtentCache() = default;
+
+    static void SetOption(const ExtentCacheOption& option);
+
+    // build extent cache from inode
+    void Build(
+        const google::protobuf::Map<uint64_t,
+                                    curvefs::metaserver::VolumeExtentList>&
+            fromInode);
+
+    void DivideForWrite(uint64_t offset,
+                        uint64_t len,
+                        const char* data,
+                        std::vector<WritePart>* allocated,
+                        std::vector<AllocPart>* needAlloc);
+
+    void DivideForRead(uint64_t offset,
+                       uint64_t len,
+                       char* data,
+                       std::vector<ReadPart>* reads,
+                       std::vector<ReadPart>* holes);
+
+    void Merge(uint64_t loffset, const PExtent& pExt);
+
+    void MarkWritten(uint64_t offset, uint64_t len);
+
+    google::protobuf::Map<uint64_t, curvefs::metaserver::VolumeExtentList>
+    ToInodePb() const;
+
+    std::unordered_map<uint64_t, std::map<uint64_t, PExtent>>
+    GetExtentsForTesting() const;
+
+ private:
+    void DivideForWriteWithinRange(const std::map<uint64_t, PExtent>& range,
+                                   uint64_t offset,
+                                   uint64_t len,
+                                   const char* data,
+                                   std::vector<WritePart>* allocated,
+                                   std::vector<AllocPart>* needAlloc);
+
+    void DivideForReadWithinRange(const std::map<uint64_t, PExtent>& range,
+                                  uint64_t offset,
+                                  uint64_t len,
+                                  char* data,
+                                  std::vector<ReadPart>* reads,
+                                  std::vector<ReadPart>* holes);
+
+    void MergeWithinRange(std::map<uint64_t, PExtent>* range,
+                          uint64_t loffset,
+                          const PExtent& extent);
+
+    void MarkWrittenWithinRange(std::map<uint64_t, PExtent>* range,
+                                uint64_t offset,
+                                uint64_t len);
+
+ private:
+    mutable curve::common::RWLock lock_;
+
+    // key is offset
+    std::unordered_map<uint64_t, std::map<uint64_t, PExtent>> extents_;
```

Comment:
extents_ needs to be limited in size or length

Reply:
nope, it's the latest extents of an inode.